### PR TITLE
[CELEBORN-1760] OOM causes disk buffer unable to be released

### PR DESCRIPTION
**Pull request by leixm**
### What changes were proposed in this pull request?
When OOM occurs in flushBuffer.addComponent, there are two problems. 

1. decrementPendingWrites is not called, causing close PartitionDataWriter to be stuck for a period of time during commit.
2. After OOM occurs, ByteBuf is not released, causing memory leaks.

### Why are the changes needed?
Fix disk buffer unable to be released

### Does this PR introduce _any_ user-facing change?
Yes, it fixes a memory leak issue in some corner cases.

### How was this patch tested?

I did some tests. 2 of the nodes did not have this PR, and the memory of these two nodes could not be released. 1 node had this PR, and the memory could be released. It was obviously much lower than the previous 2 nodes.
image

### CODE REVIEW

**Comment by leixm** [comment]
```
24/11/27 01:44:36,396 WARN [push-server-6-14] TransportChannelHandler: Exception in connection from /10.217.150.42:27112
io.netty.util.internal.OutOfDirectMemoryError: failed to allocate 4194304 byte(s) of direct memory (used: 10733223943, max: 10737418240)
        at io.netty.util.internal.PlatformDependent.incrementMemoryCounter(PlatformDependent.java:843)
        at io.netty.util.internal.PlatformDependent.allocateDirectNoCleaner(PlatformDependent.java:772)
        at io.netty.buffer.PoolArena$DirectArena.allocateDirect(PoolArena.java:717)
        at io.netty.buffer.PoolArena$DirectArena.newChunk(PoolArena.java:692)
        at io.netty.buffer.PoolArena.allocateNormal(PoolArena.java:215)
        at io.netty.buffer.PoolArena.tcacheAllocateNormal(PoolArena.java:197)
        at io.netty.buffer.PoolArena.allocate(PoolArena.java:139)
        at io.netty.buffer.PoolArena.allocate(PoolArena.java:129)
        at io.netty.buffer.PooledByteBufAllocator.newDirectBuffer(PooledByteBufAllocator.java:395)
        at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:188)
        at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:179)
        at io.netty.buffer.CompositeByteBuf.allocBuffer(CompositeByteBuf.java:1879)
        at io.netty.buffer.CompositeByteBuf.consolidate0(CompositeByteBuf.java:1758)
        at io.netty.buffer.CompositeByteBuf.consolidateIfNeeded(CompositeByteBuf.java:571)
        at io.netty.buffer.CompositeByteBuf.addComponent(CompositeByteBuf.java:266)
        at io.netty.buffer.CompositeByteBuf.addComponent(CompositeByteBuf.java:222)
        at org.apache.celeborn.service.deploy.worker.storage.FileWriter.write(FileWriter.java:224)
        at org.apache.celeborn.service.deploy.worker.PushDataHandler.writeData$1(PushDataHandler.scala:1235)
        at org.apache.celeborn.service.deploy.worker.PushDataHandler.writeLocalData(PushDataHandler.scala:1278)
        at org.apache.celeborn.service.deploy.worker.PushDataHandler.handlePushMergedData(PushDataHandler.scala:636)
        at org.apache.celeborn.service.deploy.worker.PushDataHandler.$anonfun$receive$2(PushDataHandler.scala:146)
        at org.apache.celeborn.service.deploy.worker.PushDataHandler.handleCore(PushDataHandler.scala:743)
        at org.apache.celeborn.service.deploy.worker.PushDataHandler.receive(PushDataHandler.scala:147)
        at org.apache.celeborn.common.network.server.TransportRequestHandler.processOtherMessages(TransportRequestHandler.java:132)
        at org.apache.celeborn.common.network.server.TransportRequestHandler.handle(TransportRequestHandler.java:88)
        at org.apache.celeborn.common.network.server.TransportChannelHandler.channelRead(TransportChannelHandler.java:151)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
        at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
        at org.apache.celeborn.common.network.util.TransportFrameDecoder.channelRead(TransportFrameDecoder.java:74)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:748)
```


```
24/11/27 03:05:19,880 WARN [worker-memory-manager-actor] StorageManager: Skip flushOnMemoryPressure because LocalFlusher@93798665-/mnt/ssd/0 has error: Wait pending actions timeout.
24/11/27 03:05:19,881 WARN [worker-memory-manager-actor] StorageManager: Skip flushOnMemoryPressure because LocalFlusher@93798665-/mnt/ssd/0 has error: Wait pending actions timeout.
24/11/27 03:05:19,881 WARN [worker-memory-manager-actor] StorageManager: Skip flushOnMemoryPressure because LocalFlusher@93798665-/mnt/ssd/0 has error: Wait pending actions timeout.
24/11/27 03:05:19,881 WARN [worker-memory-manager-actor] StorageManager: Skip flushOnMemoryPressure because LocalFlusher@93798665-/mnt/ssd/0 has error: Wait pending actions timeout.
24/11/27 03:05:19,881 WARN [worker-memory-manager-actor] StorageManager: Skip flushOnMemoryPressure because LocalFlusher@93798665-/mnt/ssd/0 has error: Wait pending actions timeout.
24/11/27 03:05:19,881 WARN [worker-memory-manager-actor] StorageManager: Skip flushOnMemoryPressure because LocalFlusher@93798665-/mnt/ssd/0 has error: Wait pending actions timeout
```

image

**Comment by leixm** [comment]
@RexXiong  @AngersZhuuuu  PTAL.

**Comment by FMX** [comment]
@leixm Can you provide some memory metrics with this PR after a worker is OOM?

**Comment by leixm** [comment]
```
24/11/27 01:43:22,290 INFO [worker-memory-manager-reporter] MemoryManager: Direct memory usage: 2.2 GiB/10.0 GiB, disk buffer size: 651.4 MiB, sort memory size: 0.0 B, read buffer size: 0.0 B
24/11/27 01:43:32,290 INFO [worker-memory-manager-reporter] MemoryManager: Direct memory usage: 2.5 GiB/10.0 GiB, disk buffer size: 287.9 MiB, sort memory size: 0.0 B, read buffer size: 0.0 B
24/11/27 01:43:42,290 INFO [worker-memory-manager-reporter] MemoryManager: Direct memory usage: 2.3 GiB/10.0 GiB, disk buffer size: 299.9 MiB, sort memory size: 0.0 B, read buffer size: 0.0 B
24/11/27 01:43:52,290 INFO [worker-memory-manager-reporter] MemoryManager: Direct memory usage: 3.7 GiB/10.0 GiB, disk buffer size: 272.4 MiB, sort memory size: 0.0 B, read buffer size: 0.0 B
24/11/27 01:44:02,453 INFO [worker-memory-manager-reporter] MemoryManager: Direct memory usage: 3.2 GiB/10.0 GiB, disk buffer size: 284.2 MiB, sort memory size: 0.0 B, read buffer size: 0.0 B
24/11/27 01:44:13,184 INFO [worker-memory-manager-reporter] MemoryManager: Direct memory usage: 5.5 GiB/10.0 GiB, disk buffer size: 1998.8 MiB, sort memory size: 0.0 B, read buffer size: 0.0 B
24/11/27 01:44:23,185 INFO [worker-memory-manager-reporter] MemoryManager: Direct memory usage: 4.8 GiB/10.0 GiB, disk buffer size: 1269.2 MiB, sort memory size: 0.0 B, read buffer size: 0.0 B
24/11/27 01:44:33,418 INFO [worker-memory-manager-reporter] MemoryManager: Direct memory usage: 10.0 GiB/10.0 GiB, disk buffer size: 6.8 GiB, sort memory size: 0.0 B, read buffer size: 0.0 B
24/11/27 01:44:44,008 INFO [worker-memory-manager-reporter] MemoryManager: Direct memory usage: 9.2 GiB/10.0 GiB, disk buffer size: 0.0 B, sort memory size: 0.0 B, read buffer size: 0.0 B
24/11/27 01:44:54,008 INFO [worker-memory-manager-reporter] MemoryManager: Direct memory usage: 9.2 GiB/10.0 GiB, disk buffer size: 0.0 B, sort memory size: 0.0 B, read buffer size: 0.0 B
24/11/27 01:45:04,008 INFO [worker-memory-manager-reporter] MemoryManager: Direct memory usage: 9.2 GiB/10.0 GiB, disk buffer size: 0.0 B, sort memory size: 0.0 B, read buffer size: 0.0 B
```

**Comment by leixm** [comment]
The catch Exception here will cause fileWriter.decrementPendingWrites() to not be called, but I am not sure whether it will indirectly cause the off-heap memory to not be released. I will investigate further.

**Review by turboFei — APPROVED** [review]
LGTM, should we check other exception match in the code?

**Comment by FMX** [comment]
The off-heap memory is not released if the disk buffer counter is decreased. This PR failed to complete its purpose. There is more to investigate to find out what happened before merging into main.

**Review by FMX — CHANGES_REQUESTED** [review]
More information is needed.

File: worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java [hunk 1]
```diff
@@ -416,7 +416,13 @@ public void write(ByteBuf data) throws IOException {
       }
 
       data.retain();
-      flushBuffer.addComponent(true, data);
+      try {
+        flushBuffer.addComponent(true, data);
+      } catch (OutOfMemoryError oom) {
```
Comment (leixm) [inline worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java:421]: May be Throwable will be better?

**Comment by leixm** [comment]
There are three problems in total
1. PartitionDataWriter#write flushBuffer.addComponent(true, data) After OOM occurs, data is not released.
2. After OOM occurs in PushDataHandler#writeLocalData fileWriter.write(body), fileWriter.decrementPendingWrites() is not called, which will cause some fileWriters to wait for a period of time when call close().
3. For the failed shuffle, commit was not called and the PartitionDataWriter was not closed, so there was no returnBuffer. The PartitionDataWriter should be closed when the shuffle expires.

**Comment by leixm** [comment]
@FMX  @turboFei  @RexXiong  Can you help review again?

File: worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java (hunk 1 above)
Comment (FMX) [reply worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java:421]: This makes no sense because the flush buffer has some shuffle data that was received earlier, and then OOM happens. The shuffle data is lost even after committing files.
This would cause data correctness problems here. 
That's why we never try to resume Celeborn workers after OOM happens.

Comment (leixm) [reply worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java:421]: Why the shuffle data would be lost?

Comment (FMX) [reply worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java:421]: After some investigation, there won't be shuffle data lost.

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala [hunk 2]
```diff
@@ -742,7 +742,12 @@ private[celeborn] class Worker(
     synchronized {
       val expiredApplicationIds = new JHashSet[String]()
       expiredShuffleKeys.asScala.foreach { shuffleKey =>
-        partitionLocationInfo.removeShuffle(shuffleKey)
+        partitionLocationInfo.removeShuffle(shuffleKey).foreach { loc =>
```
Comment (FMX) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/Worker.scala:745]: StorageManager will remove expired shuffle keys which will destroy the writerd and return buffer.

**Review by FMX — APPROVED** [review]
LGTM

**Review by AngersZhuuuu — APPROVED** [review]
LGTM

**Comment by pan3793** [comment]
@leixm would you mind updating the PR description to summarize the conclusion?

File: worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java [hunk 3]
```diff
@@ -416,7 +416,13 @@ public void write(ByteBuf data) throws IOException {
       }
 
       data.retain();
-      flushBuffer.addComponent(true, data);
+      try {
+        flushBuffer.addComponent(true, data);
```
Comment (RexXiong) [inline worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java:420]: Only the non-memory shuffle file storage type requires a call to releaseDiskBuffer?

**Review by RexXiong — APPROVED** [review]

**Comment by leixm** [comment]
> @leixm would you mind updating the PR description to summarize the conclusion?

I have updated the description.

**Comment by leixm** [comment]
> @leixm would you mind updating the PR description to summarize the conclusion?

Done,

**Review by pan3793 — APPROVED** [review]

**Comment by AngersZhuuuu** [comment]
ping  @FMX

**Comment by RexXiong** [comment]
Thanks merge to main(v0.6.0) and branch-0.5(v0.5.3)
